services/appflowy-collaborate/src/group/cmd.rs

```diff
@@ -169,6 +182,40 @@ where
     Ok(())
   }
 
+  /// similar to `handle_client_collab_message`, but the messages are sent from the server instead.
+  #[instrument(level = "trace", skip_all)]
+  async fn handle_server_collab_messages(
+    &self,
+    user: &RealtimeUser,
+    object_id: String,
+    messages: Vec<ClientCollabMessage>,
+  ) -> Result<(), RealtimeError> {
+    if messages.is_empty() {
+      warn!("Unexpected empty collab messages sent from server");
+      return Ok(());
+    }
+
+    if let Some(group) = self.group_manager.get_group(&object_id).await {
+      let (collab_message_sender, _collab_message_receiver) = futures::channel::mpsc::channel(1);
+      let (mut message_by_oid_sender, message_by_oid_receiver) = futures::channel::mpsc::channel(1);
```

Comment:
Use `tokio::sync::mpsc::channel`: tokio is often more mature than futures. If generic constraints allow that you could also use `tokio::sync::oneshot`.

Reply:
i have tried. but based on the `subscribe` function signature, it's not possible to use the tokio channel

```rust
  pub async fn subscribe<Sink, Stream>(
    &self,
    user: &RealtimeUser,
    subscriber_origin: CollabOrigin,
    sink: Sink,
    stream: Stream,
  ) where
    Sink: SinkExt<CollabMessage> + Clone + Send + Sync + Unpin + 'static,
    Stream: StreamExt<Item = MessageByObjectId> + Send + Sync + Unpin + 'static,
    <Sink as futures_util::Sink<CollabMessage>>::Error: std::error::Error + Send + Sync,
```